components/resource_control/src/channel.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{cell::RefCell, sync::Arc};
+
+use collections::HashMap;
+use crossbeam::channel::{self, RecvError, SendError, TryRecvError, TrySendError};
+use kvproto::kvrpcpb::CommandPri;
+use tikv_util::mpsc::priority_queue;
+
+use crate::{ResourceConsumeType, ResourceController};
+
+pub trait ResourceMetered {
+    // returns the msg consumption of each hash map
+    fn get_resource_consumptions(&self) -> Option<HashMap<String, u64>> {
+        None
+    }
+}
+
+pub fn bounded<T: Send + 'static>(
+    resource_ctl: Option<Arc<ResourceController>>,
+    cap: usize,
+) -> (Sender<T>, Receiver<T>) {
+    if let Some(ctl) = resource_ctl {
+        // TODO: make it bounded
+        let (tx, rx) = priority_queue::unbounded();
+        (
+            Sender::Priority {
+                resource_ctl: ctl,
+                sender: tx,
+                last_msg_group: RefCell::new(String::new()),
+            },
+            Receiver::Priority(rx),
+        )
+    } else {
+        let (tx, rx) = channel::bounded(cap);
+        (Sender::Vanilla(tx), Receiver::Vanilla(rx))
+    }
+}
+
+pub fn unbounded<T: Send + 'static>(
+    resource_ctl: Option<Arc<ResourceController>>,
+) -> (Sender<T>, Receiver<T>) {
+    if let Some(ctl) = resource_ctl {
+        let (tx, rx) = priority_queue::unbounded();
+        (
+            Sender::Priority {
+                resource_ctl: ctl,
+                sender: tx,
+                last_msg_group: RefCell::new(String::new()),
+            },
+            Receiver::Priority(rx),
+        )
+    } else {
+        let (tx, rx) = channel::unbounded();
+        (Sender::Vanilla(tx), Receiver::Vanilla(rx))
+    }
+}
+
+pub enum Sender<T: Send + 'static> {
+    Vanilla(channel::Sender<T>),
+    Priority {
+        resource_ctl: Arc<ResourceController>,
+        sender: priority_queue::Sender<T>,
+        last_msg_group: RefCell<String>,
+    },
+}
+
+impl<T: Send + 'static> Clone for Sender<T> {
+    fn clone(&self) -> Self {
+        match self {
+            Sender::Vanilla(sender) => Sender::Vanilla(sender.clone()),
+            Sender::Priority {
+                resource_ctl,
+                sender,
+                ..
+            } => Sender::Priority {
+                resource_ctl: resource_ctl.clone(),
+                sender: sender.clone(),
+                last_msg_group: RefCell::new(String::new()),
+            },
+        }
+    }
+}
+
+impl<T: Send + 'static> Sender<T> {
+    pub fn send(&self, m: T, low_bound: u64) -> Result<u64, SendError<T>> {
```

Comment:
Needs comment, what is low_bound, what is the return value. It seems `low_bound` only affects return value, and return value isn't being used anywhere?